oak_io/src/handle.rs

```diff
@@ -43,3 +44,201 @@ impl std::fmt::Debug for WriteHandle {
+///     handle: Handle,
+/// }
+/// # impl HandleVisit for Thing {
+/// #   fn visit<F: FnMut(&mut Handle)>(&mut self, mut visitor: F) -> F {
+/// #     visitor(&mut self.handle);
+/// #     visitor
+/// #   }
+/// # }
+///
+/// let mut thing = Thing { handle: 42 };
+///
+/// let handles = extract_handles(&mut thing);
+///
+/// assert_eq!(handles, vec![42]);
+/// ```
+pub fn extract_handles<T: HandleVisit>(msg: &mut T) -> Vec<Handle> {
+    let mut handles = Vec::new();
+    msg.visit(|handle: &mut Handle| {
+        handles.push(*handle);
+        *handle = oak_abi::INVALID_HANDLE;
+    });
+    handles
+}
+
+/// Inject handles into a message.
+///
+/// If the number of handles provided is not exactly equal to the number of handles needed to fill
+/// the message, an error is returned.
+///
+/// Order is significant: handles are injected in field declaration order, recursing
+/// into nested structs before moving on to the next field.
+///
+/// ```
+/// use oak_io::{
+///     handle::{inject_handles, HandleVisit},
+///     Handle,
+/// };
+///
+/// # #[derive(Debug, PartialEq)]
+/// struct Thing {
+///     handle: Handle,
+/// }
+/// # impl HandleVisit for Thing {
+/// #   fn visit<F: FnMut(&mut Handle)>(&mut self, mut visitor: F) -> F {
+/// #     visitor(&mut self.handle);
+/// #     visitor
+/// #   }
+/// # }
+///
+/// let handles = vec![42];
+/// let mut thing = Thing { handle: 0 };
+///
+/// inject_handles(&mut thing, &handles).unwrap();
+///
+/// assert_eq!(thing, Thing { handle: 42 });
+/// ```
+pub fn inject_handles<T: HandleVisit>(msg: &mut T, handles: &[Handle]) -> Result<(), OakError> {
+    let mut handles = handles.iter();
+    let mut result = Ok(());
+    msg.visit(|handle| {
+        if let Some(to_inject) = handles.next() {
+            *handle = *to_inject;
+        } else {
+            result = Err(OakError::ProtobufDecodeError(None));
+        }
+    });
+    if handles.next().is_some() {
+        result = Err(OakError::ProtobufDecodeError(None));
+    }
+    result
+}
+
+// A default implementation of the HandleVisit trait that does nothing
+#[macro_export]
+macro_rules! handle_visit_blanket_impl {
+    ($($t:ty),+) => {
+        $(
+            impl HandleVisit for $t {
+                fn visit<F: FnMut(&mut Handle)>(&mut self, visitor: F) -> F {
+                    visitor
+                }
+            }
+        )+
+    };
+}
+
+// Provide an implementation for all scalar types in Prost.
+// See: https://github.com/danburkert/prost#scalar-values
+handle_visit_blanket_impl!((), f64, f32, i32, i64, u32, u64, bool, String, Vec<u8>);
+
+// Provide an implementation for the Any type in Prost.
+handle_visit_blanket_impl!(prost_types::Any);
+
+// Provide an implementation for oak_abi type that an implementation cannot be derived for.
+// It does not contain handles, so a blanket impl is sufficient.
+handle_visit_blanket_impl!(oak_abi::proto::oak::application::ConfigMap);
+
+pub use oak_derive::HandleVisit;
```

Comment:
What is this? Does not it cause a circular dependency? Perhaps add a comment to explain what it does.

Reply:
This allows other crates to use  `#[derive(HandleVisit)]` without having to explicitly reference the `oak_derive` crate. Added a comment.

Reply:
Thanks. But there is still a circular dependency: `oak_io` is dependent on `oak_derive` and `oak_derive` is dependent on `oak_io`. Does Rust/Cargo generally allow circular dependency? 
Can `oak_derive` be merged into `oak_io`?

Reply:
Cargo does not allow circular references. But it is not really a circular reference. `oak_derive` is not dependent on `oak_io`, it is only the integration tests that depends on both `oak_derive` and `oak_io` to test derivation of the `HandleVisit` trait. Seeing that the integration tests end up being built as a separate binary package, the dependency graph is not circular.

The `oak_derive` crate could be merged into `oak_io`, but it seems to be considered standard practice in rust to keep the macro derivation in a separate crate so that other crates can still use the trait even if they don't want to use derivation. It would be possible to move the `oak_derive` integration tests into the `oak_io` package to remove the need for the dev dependency, but I am not sure this would make it less confusing as the tests would be removed from the code it is testing.

Reply:
Thanks for the clarification Conrad. 
I did not know about the practice of keeping the macro derivation in a separate package. I agree that it is better to keep the integration test close to the code it is testing.